mediatek: add support for TP-Link EX820v

michelelizzit opened the pull request:
Specifications:
SoC: MediaTek MT7986
RAM: 512MiB
Flash: SPI-NAND 128 MiB
Ethernet: 2x 2.5G 3x 1G
USB: 1x USB 3.0
Power: 12 VDC, 2.5 A

Installation through TFTP:
Note: UART pins on the board must be connected to USB TTL
      Serial Configuration 115200 8n1

- Boot the EX820v
- When "press ctrl-c or t to go to uboot cmdline" appears press "t"
- Connect to the board Ethernet port n.1
    (IPADDR: 192.168.1.1, ServerIP: 192.168.1.2)
- tftpboot <Firmware Image Name>
- bootm
- transfer the sysupgrade image to OpenWRT via HTTP
- fw_setenv tp_boot_idx 0 # This is needed so that uboot boots from ubi0 instead of ubi1
- sysupgrade <Sysupgrade Image>

mrkiko commented:
Are you expected to run the fw_setenv command after sysupgrade or before?
Thanks for your work.

michelelizzit commented:
Thank you @mrkiko for your feedback.
The fw_setenv command can also be performed before the sysupgrade, with the added benefit of not having to manually boot into the secondary partition.
I have now edited the description.

hauke commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -67,6 +67,9 @@ mediatek_setup_interfaces()
 		ucidef_set_interfaces_lan_wan "lan2 lan3 lan4" wan
 		;;
+	tplink,ex820v)
+		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4 lan6" "eth1 wan"
```
Why do you list 7 port here, but the device only has 5 ports?

michelelizzit replied:
Thank you @hauke, this is now fixed.

hauke commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/firmware/11-mt76-caldata:
```diff
@@ -34,6 +34,17 @@ case "$FIRMWARE" in
 		;;
+	tplink,ex820v)
+		mkdir /tmp/calmnt/
+		ubiattach -m 5 -d 1
```
Are these numbers stable?
The ubiattach command has these paramaters:
````
"-p, --dev-path=<path> path to MTD device node to attach\n"
"-m, --mtdn=<number>   MTD device number to attach (alternative method, e.g\n"
"                      if the character device node does not exist)\n"
`````
Is it possible to use `-p` instead of `-m`?

hauke replied:
did you look into this?

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/gpio/gpio.h>
+
+#include "mt7986a-rfb.dtsi"
```
Please use `mt7986a.dtsi` as template.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+#include "mt7986a-rfb.dtsi"
+
+/ {
+	compatible = "tplink,ex820v", "mediatek,mt7986a-rfb-snand";
```
```suggestion
	compatible = "tplink,ex820v", "mediatek,mt7986a";
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+		stdout-path = "serial0:115200n8";
+	};
+
+	memory {
```
```suggestion
	memory@0x40000000 {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+	};
+
+	memory {
+		reg = <0 0x40000000 0 0x10000000>;
```
```suggestion
		reg = <0 0x40000000 0 0x20000000>;
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+		compatible = "gpio-leds";
+
+	 	led-0 {
+			label = "green:lan2";
```
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_LAN;
			function-enumerator = <2>;
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+		};
+
+		led-1 {
+			label = "green:lan1";
```
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_LAN;
			function-enumerator = <1>;
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+		};
+
+		led-2 {
+			label = "green:lan0";
```
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_LAN;
			function-enumerator = <0>;
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+		};
+
+		led-3 {
+			label = "green:wan";
```
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_WAN;
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+		};
+
+		led-4 {
+			label = "orange:status";
```
```suggestion
			color = <LED_COLOR_ID_ORANGE>;
			function = LED_FUNCTION_STATUS;
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+		};
+
+		led_status_green: led-5 {
+			label = "green:status";
```
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_STATUS;
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+	pinctrl-0 = <&spi_flash_pins>;
+	status = "okay";
+
+	spi_nand_flash: flash@0 {
```
```suggestion
	flash@0 {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+				read-only;
+			};
+
+			ubenv: partition@200000 {
```
```suggestion
			partition@200000 {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+			misc_ro: partition@300000 {
+				label = "misc_ro";
+				reg = <0x300000 0x600000>;
+				// phandle = <0x20>;
```
Remove comment.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+				// phandle = <0x20>;
+			};
+
+			misc_rw: partition@900000 {
```
```suggestion
			partition@900000 {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+				reg = <0x5f00000 0x600000>;
+			};
+
+			bflag: partition@6500000 {
```
```suggestion
			partition@6500000 {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+				reg = <0x6500000 0x600000>;
+			};
+
+			misc_isp: partition@6b00000 {
```
```suggestion
			partition@6b00000 {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+				reg = <0x6b00000 0x600000>;
+			};
+
+			partition@w {
```
```suggestion
			partition@0 {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+	pinctrl-names = "default";
+	pinctrl-0 = <&wf_2g_5g_pins>;
+	
+	mediatek,mtd-eeprom = <&misc_ro 0>;
```
It's better to use `"eeprom"` nvmem cell instead.

adripo commented:
Nice work @michelelizzit ! Did you have time to look into DragonBluep's suggestions?

garydevenay commented:
Would be interested in testing this if these suggestions are being implemented.

BuzzBumbleBee commented:
I have picked up off the back of this patch set. 

The LAN side 2.5Gbps port now works, I'll submit an updated patch set in the next day or so.

@garydevenay if your still interested in testing (and have the ability to access the uarts) would you be able to get dump of the stock rootfs / kernel ?

MolecularAudio commented:
Any news on when this router will be supported? I have one which I am trying to use as a repeater - but the ISD customised firmware it comes with does not support this feature :(

abajk commented on target/linux/mediatek/dts/mt7986a-tplink-ex820v.dts:
```diff
@@ -0,0 +1,275 @@
+		#size-cells = <1>;
+		reg = <0>;
+
+		spi-max-frequency = <20000000>;
```
On most boards 52MHz works fine. Can it go higher?

michelelizzit committed:
mediatek: add support for TP-Link EX820v
Specifications:
SoC: MediaTek MT7986
RAM: 512MiB
Flash: SPI-NAND 128 MiB
Ethernet: 2x 2.5G 3x 1G
USB: 1x USB 3.0
Power: 12 VDC, 2.5 A

Signed-off-by: Michele Lizzit <michele@lizzit.it>

michelelizzit commented:
I believe I have now addressed all suggestions from DragonBluep and abajk.
All 2.5Gbps ports are now working. Most but not all LEDs are working
Thank you also BuzzBumbleBee; my fix to get it working was very similar.
I believe the only issue that remains is how to reliably extract calibration data from its ubifs partition, which was highlighted by hauke; however I am not sure I will be able to take care of that in the near future. Any help would be highly appreciated

MrBrownman commented:
Is this working and supported now?
